sdk/attestation/azure-security-attestation/test/ut/policygetset_test.cpp

```diff
@@ -0,0 +1,432 @@
+          m_endpoint, credential, options);
+    }
+
+    bool ValidateSetPolicyResponse(
+        std::unique_ptr<AttestationAdministrationClient> const& client,
+        Response<AttestationToken<PolicyResult>> const& result,
+        Azure::Nullable<std::string> policyToValidate,
+        Azure::Nullable<AttestationSigningKey> const& signingKey
+        = Azure::Nullable<AttestationSigningKey>())
+    {
+      EXPECT_EQ(result.RawResponse->GetStatusCode(), Azure::Core::Http::HttpStatusCode::Ok);
+
+      // SetPolicy responses should have updated or reset the policy value.
+      EXPECT_TRUE(result.Value.Body.PolicyResolution);
+      if (policyToValidate)
+      {
+        EXPECT_EQ(PolicyModification::Updated, result.Value.Body.PolicyResolution.Value());
+
+        // The attestation service only returns the PolicySigner and PolicySigningHash on
+        // SetPolicy calls, not ResetPolicy calls.
+
+        // Now check the policy signer if appropriate.
+        if (signingKey)
+        {
+          EXPECT_TRUE(result.Value.Body.PolicySigner);
+          EXPECT_FALSE(result.Value.Body.PolicySigner.Value().CertificateChain.Value().empty());
+
+          // When the test case type is secured, playback mode, the signing certificate was the
+          // certificate retrieved at the time the recordings were made, and it will *not* match
+          // the dummy value provided for the recorded tests.
+          if (!m_testContext.IsPlaybackMode())
+          {
+            auto signerCertificate = Cryptography::ImportX509Certificate(
+                result.Value.Body.PolicySigner.Value().CertificateChain.Value()[0]);
+            auto expectedCertificate
+                = Cryptography::ImportX509Certificate(signingKey.Value().PemEncodedX509Certificate);
+            EXPECT_EQ(expectedCertificate->GetThumbprint(), signerCertificate->GetThumbprint());
+          }
+        }
+        else
+        {
+          EXPECT_FALSE(result.Value.Body.PolicySigner);
+        }
+
+        EXPECT_TRUE(result.Value.Body.PolicyTokenHash);
+
+        // The returned PolicyTokenHash value is the hash of the entire policy JWS that was sent
+        // to the service. In playback mode, the JWS which is calculated for the tests is
+        // different from the JWS which was recorded (because the signing certificate is
+        // different).
+        //
+        // So skip verifying the PolicyTokenHash in playback mode.
+        if (!m_testContext.IsPlaybackMode())
+        {
+          AttestationToken<std::nullptr_t> sentToken
+              = client->CreateSetAttestationPolicyToken(policyToValidate, signingKey);
+
+          Azure::Core::Cryptography::_internal::Sha256Hash hasher;
+          std::vector<uint8_t> rawTokenHash = hasher.Final(
+              reinterpret_cast<const uint8_t*>(sentToken.RawToken.data()),
+              sentToken.RawToken.size());
+          EXPECT_EQ(result.Value.Body.PolicyTokenHash.Value(), rawTokenHash);
+        }
```

Comment:
Consider adding a function called `IsTokenEquals()` to the test-fw and call it here instead of `EXPECT_EQ(result.Value.Body.PolicyTokenHash.Value(), rawTokenHash);` as `isTokenEquals(result.Value.Body.PolicyTokenHash.Value(), rawTokenHash)` and let the test-fw to check if it is Playback mode.

This way we can remove the nested logic and make the test a little ignorant about the test-mode.  


```cpp
void IsTokenEquals(a, b) {
  if (!m_testContext.IsPlaybackMode()) {
     EXPECT_EQ(a, b);
  }
```

Reply:
The name can be better defined. It works on top of ASSERT_EQ for a specific type.  Its intention is to hide the test mode.

In a world where tests cases doesn't know anything about test-modes, it becomes easy to add/remove/update test-modes without adjusting any test case.  That's the beauty of a nice modular structure.

Right now, if the test-fw decides to rename test-mode. Or makes changes, it would probably break some of the Attestation test case.  Hopefully that's never the case, but I don't like when components are build and created under assumptions of things that are unlikely to happen, because when they happen, someone will have to deal with it

Reply:
Here's what I *could* see doing, but all I'm doing is shuffling the deck chairs around - the tests are *still* sensitive to the test mode, it's just that how they find the test mode changes:

```c++
void TestOnPlaybackMode(std::function<void> const& testToRun)
{
  if (m_testContext.IsPlaybackMode())
  {
    testToRun();
  }
}
```
And similar `RunTestOnLiveService` etc.

But again, all this does is move the "IsPlaybackMode" call into a test base function and doesn't change what has to happen when the test modes change.

However I'm still trying to wrap my head around this construct of "tests shouldn't know what mode in which they're running". None of the 4 languages I'm familiar with have had this construct and I'm really having a hard time understanding why it's a bad thing that tests have different conditions based on the compromises made for recordings (specifically that the environment variables retrieved in recordings are synthetic and will 100% not align with the data returned by the service). 

If you have a service which returns information based on the environment variables (like a service which echos back the endpoint in a response), or if you have a client which requires that information retrieved from environment variables be in a specific form (like a PEM encoded X.509 certificate), I don't understand why the tests shouldn't special case those checks.

However it doesn't change the fact that the checks are incredibly valuable for live tests (and for some of my code, I have entire scenarios that can only be executed when run against the service). I strongly feel that these scenarios SHOULD have test coverage, even if they're not in the CI framework.

For the isolated tests, I thought about eliminating them from the GetTestInputs function, but that's a static function - it doesn't have access to the test mode.